### PR TITLE
BlackoilAquiferModel: add serialization of dynamic state

### PR DESCRIPTION
**Pull request by akva2**
Sits on top of https://github.com/OPM/opm-simulators/pull/4425
Downstream of https://github.com/OPM/opm-common/pull/3388

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this opm-models=780 opm-common=3388 please

**Comment by akva2** [comment]
note the test part is slightly suboptimal. i ran out of patience trying to grok the parametric macro to account for the different constructor signature for AquiferNumerical.

**Comment by akva2** [comment]
jenkins build this please

File: tests/test_RestartSerialization.cpp [hunk 1]
```diff
@@ -134,6 +134,64 @@ BOOST_AUTO_TEST_CASE(EclGenericProblem)
...
+
+struct AquiferFixture {
+    AquiferFixture() {
+        using TT = Opm::Properties::TTag::EbosTypeTag;
+        const char* argv[] = {
+            "test_RestartSerialization",
+            "--ecl-deck-file-name=GLIFT1.DATA"
+        };
+        Opm::setupParameters_<TT>(2, argv, /*registerParams=*/true);
+        Opm::EclGenericVanguard::setCommunication(std::make_unique<Opm::Parallel::Communication>());
```
Comment (bska) [inline tests/test_RestartSerialization.cpp:147]: Does `Parallel::Communication` work in sequential mode?  If not, we should probably protect this with `HAVE_MPI` or similar.

File: opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp [hunk 2]
```diff
@@ -184,4 +184,23 @@ data::Aquifers BlackoilAquiferModel<TypeTag>::aquiferData() const
...
+        auto* ct = dynamic_cast<AquiferCarterTracy<TypeTag>*>(aiPtr.get());
+        auto* fetp = dynamic_cast<AquiferFetkovich<TypeTag>*>(aiPtr.get());
+        auto* num = dynamic_cast<AquiferNumerical<TypeTag>*>(aiPtr.get());
+        if (ct) {
+            serializer(*ct);
+        } else if (fetp) {
+            serializer(*fetp);
+        } else if (num) {
+            serializer(*num);
+        }
```
Comment (bska) [inline opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp:202]: This should probably blow up if neither of the `dynamic_cast<>`s succeed&mdash;e.g., if we get a new aquifer type and we forget to add support for it here.

**Review by bska — COMMENTED** [review]
I think this is okay, but I've forgotten enough about `Parallel::Communication` to be completely sure that this works in sequential mode.  Please elaborate.

File: tests/test_RestartSerialization.cpp (hunk 1 above)
Comment (akva2) [reply tests/test_RestartSerialization.cpp:147]: It's handled by a dummy in dune so it works fine (mostly). The issue we have encountered sometimes is that the dummy communication is not castable to Dune::No_Comm, while the MPI communicator is castable to MPI_Comm. In this case this is not relevant so no ifdef needed.

File: opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp (hunk 2 above)
Comment (akva2) [reply opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp:204]: yeah, added an assertion.

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp [hunk 3]
```diff
@@ -184,4 +184,25 @@ data::Aquifers BlackoilAquiferModel<TypeTag>::aquiferData() const
...
+        auto* fetp = dynamic_cast<AquiferFetkovich<TypeTag>*>(aiPtr.get());
+        auto* num = dynamic_cast<AquiferNumerical<TypeTag>*>(aiPtr.get());
+        if (ct) {
+            serializer(*ct);
+        } else if (fetp) {
+            serializer(*fetp);
+        } else if (num) {
+            serializer(*num);
+        } else {
+            assert(0);
```
Comment (bska) [inline opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp:203]: While we don't do it right now, I believe there is ongoing work to re-enable `NDEBUG` in Release mode&mdash;i.e., to [disable all `assert()`s](https://github.com/OPM/opm-common/blob/63bfa8e47cbc1742b20a5b513ffcbe10658ca6f1/cmake/Modules/UseOptimization.cmake#L66-L69)  when building in Release mode.  In that case, we probably want more protection than just `assert(false)` here.

**Comment by akva2** [comment]
that being said, you made me nervous, so let's just make sure ;)

**Comment by akva2** [comment]
jenkins build this serial please

**Review by bska — COMMENTED** [review]
Thanks for the updates.  With the possible caveat about `assert(false)`&mdash;and the sequential mode build check succeeding&mdash;I think this is ready for merging now.

File: opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp (hunk 3 above)
Comment (akva2) [reply opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp:203]: hmm. i'm not a fan of throwing due to logic/programming errors, but i see you point. exception it is then.

**Comment by akva2** [comment]
serial build was green.

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now.  Please merge when the current build check is green.

**Pull request merged**